File: serie4/musterloesung/acker.m
MODULE ackermann;

VAR
    x : INTEGER;
    a : INTEGER;
    b : INTEGER;

PROCEDURE acker(VAR x : INTEGER; VAR y : INTEGER) : INTEGER;
BEGIN
    IF x = 0 THEN
        RETURN y + 1
    ELSIF y = 0 THEN
        RETURN acker(x - 1, 1)
    ELSE
        RETURN acker(x - 1, acker(x, y - 1))
    END
END acker;

BEGIN
    a := 3;
    b := 2;
    x := acker(a,b)
END ackermann.

File: serie5/acker.c
#include <stdio.h>
/* compiled from module "ackermann" */ 

/* global declarations */
int x;
int a;
int b;

int acker(int x, int y) {
/* acker's local declarations */

/* code */
if ((x == 0)) {
return (y + 1);
} else if ((y == 0)) {
return acker((x - 1), 1);
} else {
return acker((x - 1), acker(x, (y - 1)));
};
}


/* main program */
int main() {
a = 3;
b = 2;
x = acker(a, b);
printf("x = %d\n" ,x);
}
